components/sup/src/manager/service.rs

```diff
@@ -168,10 +169,110 @@ impl TemplateUpdate {
+    ticks:            u64,
+    restart_attempts: u64,
+    restart_state:    RestartState,
+    restart_backoff:  Backoff,
+}
+
+impl PersistentServiceWrapper {
+    pub fn new(service: Service) -> PersistentServiceWrapper {
+        PersistentServiceWrapper { run_state: ServiceRunState::default(),
+                                   inner:     Some(service), }
+    }
+
+    pub fn service(&self) -> Option<&Service> { self.inner.as_ref() }
+
+    pub fn service_mut(&mut self) -> Option<&mut Service> { self.inner.as_mut() }
+
+    pub fn start(&mut self, service: Service) {
+        self.run_state.restart_state = match self.run_state.restart_state {
+            RestartState::None => RestartState::None,
+            RestartState::NeedsRestart => {
+                panic!("Start called on service which was not ready to be restarted");
+            }
+            RestartState::Restarting => {
+                outputln!(preamble service.service_group, "Restarted");
+                RestartState::Restarted
+            }
+            RestartState::Restarted => {
+                panic!("Start called on service which was already restarted")
+            }
+        };
+        self.inner = Some(service);
+    }
+
+    pub fn shutdown(&mut self, is_restart: bool) -> Option<Service> {
+        if let Some(service) = &self.inner {
+            if is_restart {
+                let restart_duration = self.run_state
+                                           .restart_backoff
+                                           .record_attempt()
+                                           .unwrap_or_default();
+                self.run_state.restart_attempts += 1;
+                self.run_state.restart_state = match self.run_state.restart_state {
+                    RestartState::None => RestartState::Restarting,
+                    RestartState::NeedsRestart => RestartState::Restarting,
+                    RestartState::Restarting => {
+                        panic!("Shutdown called on service which was already restarting")
+                    }
+                    RestartState::Restarted => RestartState::Restarting,
+                };
+                outputln!(preamble service.service_group, "Stopping service, will restart after {:.2} secs", restart_duration.as_secs_f32());
+            }
+            self.inner.take()
+        } else {
+            None
+        }
+    }
+
+    pub fn restart_state(&self) -> RestartState { self.run_state.restart_state }
+
+    pub fn is_ready_for_restart(&self) -> bool {
+        self.run_state.restart_state == RestartState::Restarting
+        && self.run_state
+               .restart_backoff
+               .duration_until_next_attempt()
+               .is_none()
+    }
+
+    pub fn tick(&mut self, census_ring: &CensusRing, launcher: &LauncherCli) -> bool {
+        match &mut self.inner {
+            Some(ref mut service) => {
+                debug!("Starting service tick with persistent state: {:?}",
```

Comment:
I'd be inclined to make this a `trace!` since it is called so frequently

Reply:
Agreed